### DISCUSSION TITLE
### 📣 GitHub’s Customer Engagement team want to hear from you

### DISCUSSION THREAD

**Discussion by juschadha** [community/community#27366] [Other Feature Feedback, Questions, & Ideas]
We are keen to learn more about how you chose and implemented GitHub in your organization.
Please share any tips/best practices (1-2) that helped you drive usage and adoption?

**Comment by MrSarno** [comment]
_(The below technically applies to a previous workplace.)_

----

We primarily used the following features of GitHub;

- Code repositories

  - Issues
  - Pull requests
  
- Projects (Classic)

  - Boards

Features I believe could have been helpful but we did not use (much);

- Code repositories

  - Discussions
  - Wikis

----

The context for this was an indie game development studio using the Unity game engine & needing a place to store code, track bugs, and use kanban-style boards for agile development. It made sense to have as much of this as possible on a single platform.

Had discussions always been around, I think we might have gotten a lot of use out of them. As it was, we used Discord as a free alternative to Slack for real-time chat, and we used Confluence for documentation we wished to maintain. Discussions sort of felt like an awkward middle-ground - not a true live chat service, but also not something we would seek to go back and edit outdated information to keep it up-to-date. I would highlight Stack Overflow for Teams as a possible competitor that I think would be better able to carve out its own niche. That said, perhaps discussions have their place & are simply best-suited to collections of open source developers that aren't all part of the same organisation - I don't want to conflate "didn't work for us" with "useless in all cases".

I don't have any specific feedback for wikis, but we ended up going with Confluence over them in _most_ cases for documentation; I think this was partly down to the perception - accurate or otherwise - that, should the documentation for one project grow very large, Confluence would scale more gracefully. I'm not sure how pleasant navigation is on GitHub wikis of vast sizes?

----

We were a SME that were remote-only before ~~Covid~~ it was cool, so we were a bit atypical at the time, but perhaps less so now.

I'm not sure how helpful this was - I'd be happy to answer any specific questions.

As for why we chose GitHub initially? You're big enough to be the default choice; we had no reason to go elsewhere.

**Reply by juschadha** [reply]
This is super insightful thank you for sharing!! We do have a lot more of that single platform experience.

I have found that with Wikis it really depends on how each team decides to set it up, and how detailed, combining it with issues and boards has been super powerful. Be keen to see if you are using it now.
Did you find it easy enough to drive usage and collaboration of issues, pull requests and boards with other developers? Was there any ramp-up time?

**Reply by MrSarno** [reply]
No; I can’t say I’ve had any experience with GitHub wikis since then.

I still wonder what tracking down a particular page looks like if, say, a wiki has 500 pages. Or 250. Or even 100?

I looked around at a handful of huge projects that interest me personally;

- [Linux](https://github.com/torvalds/linux) uses Markdown files in a documentation directory with no wiki enabled. To be fair, I think the repo predates wikis?
- [Golang](https://github.com/golang/go) appears to make thorough and excellent use of [a wiki](https://github.com/golang/go/wiki), with a vertical pages list for navigation. If this is some option they enabled, then excellent - I didn’t know about that. If some poor person had to manually list every page… less nice. [The docs](https://docs.github.com/en/communities/documenting-your-project-with-wikis/about-wikis) don’t suggest an automated option. I’d love to see a sidebar like this, but for categories instead of individual pages.
- [(Your) Docs](https://github.com/github/docs) seem to use a handful of MD files that link to each other. No wiki here, either.

I don’t know whether GitHub wikis are fundamentally geared towards smaller projects, or have the tools & features to scale beyond that, but don't promote them well. I feel they need both categories and ways of automatically generating navigation options which list pages and / or categories. Or, failing that, _very_ compelling alternatives. I can’t say there are none, but I know of none.

My asks may seem unreasonable, but when it comes to documentation, I never feel certain ahead of time how large a web app or desktop application may become - and I don’t want to commit to something that seems to have a ceiling. Issues, PRs, Boards, and the like don’t give me that concern… when it comes to wikis, I’d rather spin up MediaWiki or Wiki.js for their organisation tools.

——

Looking back, I would say;

- Pull Requests - no ramp up time
- Issues - up to a week or two for some; we did have to transition from regular Discord messages _very_ early on, but it went well
- Boards - initially leaned on a Producer / DevOps guy (I said we were indie! 😉), but over a couple of months we grew out of that

**Reply by juschadha** [reply]
Appreciate you sharing what works and what documentation could be helpful. Thanks for sharing those examples, it does seem it is team or project dependent on how things are maintained and managed. 

Good to know moving over to issues went well.

**Comment by Wazbat** [comment]
We're a relatively small organization, however we're currently doing our best to move our work to one central platform. The new project boards look very very promising, and we feel it's a good balance of dev and product focused features, allowing us developers to simply create pull requests, link them to an issue, and not have to worry about updating cards on a different website

We're thinking of using repository wiki's as an alternative to our current intranet, however the inability to search page content, especially across multiple org repos, might be a blocker. There's also the fact certain features like mermaid not being supported in wikis, but I understand that that is a work in progress

It's been interesting to all these new features, and the fact there are plenty of examples, entertaining videos on social channels, and blog posts, means it's easy to discover features that we weren't aware of before

**Reply by juschadha** [reply]
Thank you for sharing :). I'm so glad to hear your feedback on the new project boards, they have helped me manage multiple projects and issues in different repos. Also great to hear the videos and blogs have helped share what's new and coming.

**Comment by ng-druid** [comment]
Over at [Roll The Cloud](https://rollthecloud.com) we have ditched databases repurposed github repos and pages to act as highly available, scalable, reliable free NoSQL master storage solutions for JSON and media files with all goodness of git carried over to data. This enabled us to align with one of the core principles of [green software engineering](https://principles.green/) energy reduction through elimination of the traditional energy inefficient infrastructure to host data storage systems.

* [Objects repo](https://github.com/rollthecloudinc/ipe-objects-prod/tree/master/panelpage)
* [Object Access](https://rollthecloudinc.github.io/ipe-objects-prod/panelpage/3c3f243f-0ccc-4afc-869d-c37c361cd8fc.json)

We have further reduce energy consumption of our apps by embedding data directly into our build. When the application is visited the data is loaded into idb from memory. This eliminates the traditional necessity to make external requests to APIs to fetch data as users interact with the web experience. This would not have easily been possible without the extensive capabilities of github actions and community that supports it.

* [Build File](https://github.com/rollthecloudinc/druid/blob/master/.github/workflows/druid-workflow.yml)

[Demo Page Editor](https://demo.druidcloud.io/dev-test-virtual-list-flex-v1/manage)
<img width="1440" alt="Screen Shot 2022-08-02 at 12 14 54 AM" src="https://user-images.githubusercontent.com/73197190/182290447-d275e408-3a9c-46a0-9448-6da2c88759bc.png">

In the near future the build file will generate our [pre-render](https://angular.io/guide/prerendering) [routes file](https://github.com/rollthecloudinc/ipe/blob/master/data/routes.txt) that is currently static in each of our derivative repos.

One specific limitation we have run into is being able to host multiple pages environments from the same repo. This has lead use to creating separate static repos for our objects data stores.

* [dev](https://github.com/rollthecloudinc/ipe-objects/tree/dev)
* [prod](https://github.com/rollthecloudinc/ipe-objects-prod)

It would be ideal to host multiple pages sites from the same repo using branches instead.

The other limitation that I have to mention is being able to rewrite 404s on pages to either the index or a specific page. This has completely preventing us from using github pages for our main sites. Instead he have resorted to alternatives like Azure Static Sites which has a very [comprehensive rewrite configuration](https://docs.microsoft.com/en-us/azure/static-web-apps/configuration#fallback-routes) available. I would love to ditch azure and centralize all hosted sites on github pages instead as our [projects](https://github.com/rollthecloudinc/dc) default hosting platform.

**Reply by juschadha** [reply]
Thanks for sharing, this is great to see how you have been able to move over and use more of GitHub, as well as Actions, and to further reduce energy consumption. I'll be sharing internally as well :)

**Comment by maxandriani** [comment]
In my organization we do not use Github, instead we use Azure Devops. But I use GitHub to host my personal projects, some open source libraries and tecnology studies.

My main issue w/ github to host projects is that I can't manage properly the life cicle of issues inside Github. I know there is the (new) projects feature, that seem to look nice. But recently I tried to create a new project (and manage the issues) w/ this feature and I'm strugling a lot to setup my workflow.

Basically I'm planning to create a project w/ 3 different applications. Because there shall use different technologies I do not feel confortable to keep the entire project on the same repository. Mono Repo is a kind of anti pattern to the overall architecture i'm planning to the project.

Initially I felt happy to discover the project functionality, but as soon as I started to create some use-case issues I realize that I actually need a repository to store the project-level issues. The problem starts here. I have project-level issues (use cases and documentation activities) and application-level issues (per step by step app activities).

Long story short, I created the following repository structure where `app-docs` will play as main repository to store all use-case documents and the other will act like its children:

- app-docs
- app-server
- app-web
- app-ios
- ...

Then I decided to keep the use cases issues in the app-docs repository and use it to create the overall documentation in markdown format. That decision resulted in some problems:

- Althougth Github Projects allows me to visualize all issues in a single board, I'm not capable to easy see the relation between tasks. I have a main issue (use case description) and several child issues distributed all over the other repositories. In the board I can mannualy arrange then close to each other, but some kind of "hierarchy" representation would be very pleasunt. I'm using the taks-list functionality in the use-case issues, so, inside the main issue I can keep tracking of the overall activity evolution, but in the board view I have no clue to see the overal progress of the project.
- Because I decided the main repository will be a documentation storage. I'm planning to create a entire Markdown based files to store the architectural level documentation. Then I realize the wiki functionality actually creates another sub repository in the documentations repo. I'm not confortable to split my documentation bewteen main repository and Github Wiki. It is like an anti pattern. So it would be very nice if you guys allow us to point the wiki repository to the main repo like Azure Devops does. That way I'll be able to use Wiki's fully capabilities and keep source code and documentation in one place.
- Maybe if you guys consider the hability to create project-level issues, and project-level wiki, it should resolve a lot of my problems.

**Reply by Wazbat** [reply]
I believe you can create project board entries without converting them to issues which does make the project-level, but that does mean they do indeed lack the features an issue would have
